### PR TITLE
Add support for NuttX

### PR DESCRIPTION
**Pull request by v01d**
This PR adds support for Apache NuttX RTOS. As NuttX is POSIX compatible, I mostly used the Linux porting layer unchanged and only introduce some differences (I also wrote the event queue using POSIX message queues, which removed the need for the C++ file). Support for the required interface in NuttX has just been merged so this depends on master version of NuttX at the moment. nimBLE is used on NuttX just for its host layer.

I'm marking this as draft because I would like to get some feedback in order to check that I'm in the right direction and then clean it up a bit (there are some changes that you may want to be merged separately). Also, I needed to distinguish Linux from NuttX case in the HCI socket interface file but I don't know how to create a new variable to do so, so I would need to fix that as well.

I've been using nimBLE extensively on NuttX, both in the "sim" build (where NuttX uses the PC's bluetooth stack and simply passes the HCI messages around) and on nRF52832 chip (I wrote a link-layer for NuttX which is not yet merged).

Integration will be completed in NuttX by having an extra "nimble" application which downloads this repo and builds the required userspace component (the host-layer).

### CODE REVIEW

File: nimble/host/src/ble_hs_hci.c [hunk 1]
```diff
@@ -339,11 +339,13 @@ ble_hs_hci_cmd_tx(uint16_t opcode, const void *cmd, uint8_t cmd_len,
 static void
 ble_hs_hci_rx_ack(uint8_t *ack_ev)
 {
+#if 0
     if (ble_npl_sem_get_count(&ble_hs_hci_sem) > 0) {
         /* This ack is unexpected; ignore it. */
         ble_hci_trans_buf_free(ack_ev);
         return;
     }
+#endif
```
Comment (v01d) [inline nimble/host/src/ble_hs_hci.c:348]: This is ifdefed since it is a race condition

**Comment by v01d** [comment]
Ok, I just addressed the style issues and also found how to define a variable to distinguish NuttX from Linux. There are just two things worth mentioning:
- a check on a semaphore which is actually a race condition (see annotation above) since it checks the semaphore at a time which may be incorrect (in NuttX this was triggered incorrectly, possibly on Linux it was slower and it wasn't triggered)
- I had to disable LOG_* defines in log.h, due to conflicts of syslog() defines which NuttX has. It was suggested in #864 that I could redefine include path to my own log.h, and to do this I would have to change all includes of `log.h` from using `"` to `<`/`>` for that to work. But I feel hesitant to do so since it sounds a bit hack to do that only for that include. If you confirm it is OK to do so, I will add that as well. Otherwise, I think the simpler change is to remove those LOG_* macros which are not used in nimBLE (or maybe conditionally define them based on MYNEWT?)

File: nimble/host/services/bas/include/services/bas/ble_svc_bas.h [hunk 2]
```diff
@@ -28,4 +28,6 @@
 
 int ble_svc_bas_battery_level_set(uint8_t level);
 
+void ble_svc_bas_init(void);
```
Comment (v01d) [inline nimble/host/services/bas/include/services/bas/ble_svc_bas.h:31]: This declaration was missing

**Comment by v01d** [comment]
I tried fixing the remaining style issues but I'm not sure if you actually want me to change the alignment on that macro definition on log_common.h

**Comment by v01d** [comment]
Hi @ccollins476ad @andrzej-kaczmarek any chance you could take a look at the PR and offer some feedback?

**Comment by v01d** [comment]
Hi, sorry for insisting but is there a chance for this to be reviewed? I don't think it is really that different than the Linux case.
I also would like to continue BLE work on NuttX using nimBLE and having this integrated upstream would really help.

**Comment by andrzej-kaczmarek** [comment]
@v01d sorry it takes so long, I was quite busy recently but I should find some time to check it this week

**Comment by v01d** [comment]
great, thanks @andrzej-kaczmarek

File: nimble/host/src/ble_hs_hci.c (hunk 1 above)
Comment (sjanc) [reply nimble/host/src/ble_hs_hci.c:348]: could you comment on this a bit? it looks like this code was here since day 1

File: porting/examples/nuttx/README.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+## Overview
+
+See (https://mynewt.apache.org/network/ble/ble_intro/).
+
+## Building
+
+NimBLE is usually built as a part of Apache Mynewt OS, but ports for
+other RTOS-es are also available.
+
+### Linux
```
Comment (sjanc) [inline porting/examples/nuttx/README.md:33]: this readme should be either updated to reflect nuttx env or removed

File: porting/examples/nuttx/include/syscfg.h [hunk 4]
```diff
@@ -0,0 +1,941 @@
+/**
```
Comment (sjanc) [inline porting/examples/nuttx/include/syscfg.h:1]: is this file needed?

File: porting/nimble/include/log/log.h [hunk 5]
```diff
@@ -30,11 +30,13 @@ log_dummy(void *log, ...)
     (void)log;
 }
 
+#if 0
```
Comment (sjanc) [inline porting/nimble/include/log/log.h:33]: why is this needed?

File: porting/targets/nuttx/pkg.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+pkg.name: "porting/targets/nuttx"
+pkg.type: "target"
+pkg.description: This target is used to generate syscfg.h file and other artifacts for linux example app.
```
Comment (sjanc) [inline porting/targets/nuttx/pkg.yml:21]: "for nuttx example app"

File: nimble/transport/socket/src/ble_hci_socket.c [hunk 7]
```diff
@@ -205,6 +208,8 @@ static struct ble_hci_sock_state {
 static int s_ble_hci_device = MYNEWT_VAL(BLE_SOCK_TCP_PORT);
 #elif MYNEWT_VAL(BLE_SOCK_USE_LINUX_BLUE)
 static int s_ble_hci_device = MYNEWT_VAL(BLE_SOCK_LINUX_DEV);
+#elif MYNEWT_VAL(BLE_SOCK_USE_LINUX_BLUE) || MYNEWT_VAL(BLE_SOCK_USE_NUTTX)
```
Comment (andrzej-kaczmarek) [inline nimble/transport/socket/src/ble_hci_socket.c:211]: `MYNEWT_VAL(BLE_SOCK_USE_LINUX_BLUE)` is always false here since it's already checked earlier

File: nimble/transport/socket/src/ble_hci_socket.c [hunk 8]
```diff
@@ -281,6 +343,9 @@ ble_hci_sock_cmdevt_tx(uint8_t *hci_ev, uint8_t h4_type)
     memset(&msg, 0, sizeof(msg));
     memset(iov, 0, sizeof(iov));
 
+    msg.msg_name = (void *)btaddr;
+    msg.msg_namelen = 6;
+
```
Comment (andrzej-kaczmarek) [inline nimble/transport/socket/src/ble_hci_socket.c:348]: why do we need this here? there's separate code for nuttx below. if this fixes something else than nuttx, please create separate PR.

File: porting/npl/nuttx/src/os_callout.c [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    return c->c_active;
+}
+
+int ble_npl_callout_inited(struct ble_npl_callout *c)
+{
+    return (c->c_timer != NULL);
+}
+
+ble_npl_error_t ble_npl_callout_reset(struct ble_npl_callout *c,
+				      ble_npl_time_t ticks)
```
Comment (andrzej-kaczmarek) [inline porting/npl/nuttx/src/os_callout.c:79]: use spaces for indentation, return type should be in separate line. the same for other functions in NPL

File: porting/npl/nuttx/src/os_callout.c [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    if (ticks == 0) {
+        ticks = 1;
+    }
+
+    c->c_ticks = ble_npl_time_get() + ticks;
+
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;                     // one shot
```
Comment (andrzej-kaczmarek) [inline porting/npl/nuttx/src/os_callout.c:94]: do not use C++ style comments

**Review by andrzej-kaczmarek — CHANGES_REQUESTED** [review]
overall this looks good, few minor comments
please make sure NPL files comply with NimBLE coding style. other files that are "extras" for NuttX code (like tests) can be left as-is.

File: nimble/host/src/ble_hs_hci.c (hunk 1 above)
Comment (v01d) [reply nimble/host/src/ble_hs_hci.c:348]: I found that this check would return true unexpectedly and it would not get to the release. It seemed to be a race condition due to the fact that the this function would get called too fast under certain conditions. I think the sequence was that the lock was actually set after the if check.

File: nimble/transport/socket/src/ble_hci_socket.c (hunk 8 above)
Comment (v01d) [reply nimble/transport/socket/src/ble_hci_socket.c:348]: Right. I initially tried to use sendmsg() for NuttX but that is not implemented yet for Bluetooth sockets, this was leftover code. I just removed it.

File: porting/examples/nuttx/README.md (hunk 3 above)
Comment (v01d) [reply porting/examples/nuttx/README.md:33]: I'll remove it since the build starts on NuttX side and the documentation should be there.

File: porting/nimble/include/log/log.h (hunk 5 above)
Comment (v01d) [reply porting/nimble/include/log/log.h:33]: Please see previous comment: https://github.com/apache/mynewt-nimble/pull/878#issuecomment-718964467

File: porting/targets/nuttx/pkg.yml (hunk 6 above)
Comment (v01d) [reply porting/targets/nuttx/pkg.yml:21]: Fixed

File: porting/npl/nuttx/src/os_callout.c (hunk 9 above)
Comment (v01d) [reply porting/npl/nuttx/src/os_callout.c:79]: ok, I fixed coding style, let me know if you find other issues

File: porting/npl/nuttx/src/os_callout.c (hunk 10 above)
Comment (v01d) [reply porting/npl/nuttx/src/os_callout.c:94]: done

File: porting/examples/nuttx/include/syscfg.h (hunk 4 above)
Comment (v01d) [reply porting/examples/nuttx/include/syscfg.h:1]: No, I just removed it.

**Comment by v01d** [comment]
I just force pushed the changes. I still don't know what to do about the coding style error reported by CI.

File: nimble/host/src/ble_hs_hci.c (hunk 1 above)
Comment (v01d) [reply nimble/host/src/ble_hs_hci.c]: Ok, I looked further into the issue and it was not a race condition after all. The problem is that ble_npl_sem_get_count() returns unsigned integer and NuttX returns negative numbers in sem_getvalue() to indicate number of lock count (this is a valid option as per the man page). So this would generate 65536 as the sem value and the if would get triggered.
It now works as expected.

File: porting/nimble/include/log/log.h (hunk 5 above)
Comment (v01d) [reply porting/nimble/include/log/log.h:33]: Any suggestion on what to do about it? I think it is the only remaining issue in the PR

Comment (sjanc) [reply porting/nimble/include/log/log.h:33]: Hmm I think for now we could just put all this under #ifndef LOG_DEBUG

in long term I think we should add own NPL_LOG_* and use those instead of relying on modlog module defining those

File: porting/nimble/include/log_common/log_common.h [hunk 11]
```diff
@@ -26,14 +26,116 @@
...
+extern struct log_info g_log_info;
+
+/** @typedef log_append_cb
+ * @brief Callback that is executed each time the corresponding log is appended
+ * to.
+ *
+ * @param log                   The log that was just appended to.
+ * @param idx                   The index of newly appended log entry.
+ */
+typedef void log_append_cb(struct log *log, uint32_t idx);
```
Comment (sjanc) [inline porting/nimble/include/log_common/log_common.h:130]: I think we need to #include <stdint.h>  in this file, this should fix CI error for mesh sample

File: porting/nimble/include/log/log.h (hunk 5 above)
Comment (v01d) [reply porting/nimble/include/log/log.h:33]: I think that would require an include to be present before defining LOG_DEBUG which may not always be the case. What about `#if MYNEWT`?

File: porting/nimble/include/log_common/log_common.h (hunk 11 above)
Comment (v01d) [reply porting/nimble/include/log_common/log_common.h:130]: Ok, I'll add the include

File: porting/nimble/include/log/log.h (hunk 5 above)
Comment (andrzej-kaczmarek) [reply porting/nimble/include/log/log.h]: I think `#if MYNEWT` is ok for now, we'll need to resolve this anyway in some other way as @sjanc proposed

**Comment by v01d** [comment]
It seems last few changes made CI happy. I think those were all requested changes. Do you think it is OK to merge?

**Review by sjanc — APPROVED** [review]

**Review by andrzej-kaczmarek — APPROVED** [review]
